File: frontend/src/types/owner.ts
// Owner domain types — mirror the frozen contract (§3.4, §3.5).
import type { BaseListParams } from './common';

/** Request body for POST/PUT /api/v1/owners (§3.4). */
export interface OwnerCreate {
  full_name: string;
  email?: string | null;
  phone?: string | null;
  address?: string | null;
}

export type OwnerUpdate = OwnerCreate;

/** Single-read response — includes derived `pet_count` (§3.4). */
export interface Owner {
  id: number;
  full_name: string;
  email: string | null;
  phone: string | null;
  address: string | null;
  /** Derived count; present on single-read, omitted from list items. */
  pet_count?: number;
  created_at: string;
  updated_at: string;
}

/** List item — same shape as Owner but `pet_count` is not guaranteed. */
export type OwnerListItem = Owner;

export type OwnerSortField = 'full_name' | 'created_at';

/** Query params for GET /api/v1/owners (§3.5). */
export interface OwnerListParams extends BaseListParams {
  sort?: OwnerSortField;
}

File: backend/app/schemas/owner.py
"""Pydantic v2 schemas for Owner — request bodies and response shapes (architecture §3.4)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class OwnerCreate(BaseModel):
    full_name: str = Field(..., min_length=1, max_length=150)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=20)
    address: str | None = Field(default=None, max_length=255)


# PUT is a full-replace update; required fields match OwnerCreate.
OwnerUpdate = OwnerCreate


class OwnerListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    full_name: str
    email: str | None
    phone: str | None
    address: str | None
    created_at: datetime
    updated_at: datetime


class OwnerResponse(OwnerListItem):
    """Single-resource response — includes derived pet_count (architecture §3.4)."""

    pet_count: int
